packages-nextgen/kestrel_core/src/kestrel/ir/instructions.py

```diff
@@ -10,11 +20,66 @@
     BoolExp,
 )
 
+from kestrel.exceptions import (
+    InvalidInstruction,
+    InvalidSeralizedInstruction,
+)
 
+
+@dataclass
 class Instruction(DataClassJSONMixin):
+    id: uuid.UUID = field(init=False)
+    instruction: str = field(init=False)
+
+    def __post_init__(self):
+        # stable id during Instruction lifetime
+        self.id = uuid.uuid4()
+        self.instruction = self.__class__.__name__
+
+    def __hash__(self):
+        # stable hash during Instruction lifetime
+        return hash(self.id)
+
+
+@dataclass
+class Variable(Instruction):
+    name: str
+    deceased: bool = False
+
+
+@dataclass
+class Source(Instruction):
+    uri: str
+    interface: str
```

Comment:
Is `interface` redundant since we have `uri` already?  Seems like we would either store `interface` and `name` or just `uri` (which is of the form "<interface>://<name>").

Reply:
I try to solve it with the best solution I think: using `InitVar` in `dataclasses` for `uri`, so we can use `uri` in the constructor. However, I encountered an issue when combining `InitVar` with `mashumaro` (serialization or deserialization): https://github.com/Fatal1ty/mashumaro/issues/179

So I just create `source_from_uri()` as explicit constructor and keep the `Source` class simple for now.